### PR TITLE
Add buffer caching to minimize allocations.

### PR DESCRIPTION
**Pull request by a-feld**
This is an attempt to minimize heap allocations when the factory is expected to be long-lived. It does so by re-using the buffers between `BuildRequest` calls.

### CODE REVIEW

File: telemetry/request_factory.go [hunk 1]
```diff
@@ -97,24 +104,36 @@ func (f *requestFactory) buildRequest(entries []PayloadEntry, getPayloadBytes pa
 		}
 	}
 
-	buf := &bytes.Buffer{}
-	getPayloadBytes(buf, entries)
+	// Grab a buffer from the cached buffers and reset it
+	decompressedBuffer := configuredFactory.uncompressedBuffers.Get().(*bytes.Buffer)
```
Comment (jack-berg) [inline telemetry/request_factory.go:108]: Is this ever put back into the pool?

File: telemetry/request_factory.go [hunk 2]
```diff
@@ -97,24 +104,36 @@ func (f *requestFactory) buildRequest(entries []PayloadEntry, getPayloadBytes pa
...
+	decompressedBuffer.Reset()
+
+	// Grab a gzip structure (and buffer) from the cache and reset it
+	poolEntry := configuredFactory.zippers.Get().(*gzipPoolEntry)
```
Comment (jack-berg) [inline telemetry/request_factory.go:112]: Is there any advantage to having the gzip pool entries also carry a buffer? Why not just get two byte buffers from the `uncompressedBuffers` pool?

Comment (a-feld) [reply telemetry/request_factory.go:113]: We want to reuse the buffer after the size has been adjusted by the gzip writer. Those buffers will bias to smaller sizes, therefore they are kept separate.

File: telemetry/request_factory.go (hunk 1 above)
Comment (a-feld) [reply telemetry/request_factory.go:108]: :doh:

File: telemetry/request_factory.go (hunk 2 above)
Comment (jack-berg) [reply telemetry/request_factory.go:113]: I buy that.

File: telemetry/request_factory.go [hunk 3]
```diff
@@ -97,24 +104,37 @@ func (f *requestFactory) buildRequest(entries []PayloadEntry, getPayloadBytes pa
...
-	getPayloadBytes(buf, entries)
+	// Grab a buffer from the cached buffers and reset it
+	decompressedBuffer := configuredFactory.uncompressedBuffers.Get().(*bytes.Buffer)
+	defer configuredFactory.uncompressedBuffers.Put(decompressedBuffer)
+	decompressedBuffer.Reset()
+
+	// Grab a gzip structure (and buffer) from the cache and reset it
+	poolEntry := configuredFactory.zippers.Get().(*gzipPoolEntry)
+	defer configuredFactory.zippers.Put(poolEntry)
+	poolEntry.compressedBuffer.Reset()
```
Comment (jack-berg) [inline telemetry/request_factory.go:115]: Does it make a difference if we resent the `compressedBuffer` here vs. in a `defer`?

**Pull request closed without merging**